File: Balanced Binary Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode *root, int &depth) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (root == NULL) {
            depth = 0;
            return true;
        }
        int leftdepth, rightdepth;
        if (!helper(root->left, leftdepth) || !helper(root->right, rightdepth)) return false;
        
        depth = max(leftdepth, rightdepth) + 1;
        if (abs(leftdepth-rightdepth) > 1) { 
            return false;
        } else {
            return true;    
        }
    }
    
    bool isBalanced(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int depth;
        return helper(root, depth);
    }
};

File: Balanced Binary Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsBalanced(TreeNode root) {
        int depth;
        return IsBalanced(root, out depth);
    }
    
    private bool IsBalanced(TreeNode root, out int depth) {
        if (root == null) {
            depth = 0;
            return true;
        }
        
        int left;
        int right;
        if (!IsBalanced(root.left, out left) || !IsBalanced(root.right, out right)) {
            depth = 0; // No sense
            return false;
        }
        
        depth = Math.Max(left, right) + 1;
        return Math.Abs(left - right) <= 1;
    }
}
